bootloader: set configure_locale=1 in interpreter pre-config

rokm opened the pull request:
Have bootloader set the `configure_locale` field in the interpreter pre-config structure, so that user-preferred locale is set during interpreter pre-initialization.

This fixes regression w.r.t. PyInstaller 5.x and unfrozen python, where calling `locale.getlocale` in frozen application returns user-preferred locale, while frozen applications created by PyInstaller 6.x currently return `(None, None)`. By extension, this fixes #8305.

In PyInstaller 5.x bootloader, user-preferred locale was set explicitly by bootloader calling `setlocale` in C code; now, we can let python interpreter (pre)initialization handle that for us.

rokm committed:
bootloader: set configure_locale=1 in interpreter pre-config

Have bootloader set the `configure_locale` field in the interpreter
pre-config structure, so that user-preferred locale is set during
interpreter pre-initialization.

This fixes regression w.r.t. PyInstaller 5.x and unfrozen python,
where calling `locale.getlocale` in frozen application returns
user-preferred locale, while frozen applications created by
PyInstaller 6.x currently return `(None, None)`.

In PyInstaller 5.x bootloader, user-preferred locale was set
explicitly by bootloader calling `setlocale` in C code; now,
we can let python interpreter (pre)initialization handle that
for us.

rokm committed:
tests: add a test for user-preferred locale being set

Add a test that ensures that user-preferred locale is set on the
POSIX systems; i.e., that `locale.getlocale()` does not return
`(None, None)` and that values in the tuple correspond to
locale/encoding set via environment variable (`LC_ALL`).

bwoodsend commented on tests/functional/test_basic.py:
```diff
@@ -417,6 +417,29 @@ def test_time_module_localized(pyi_builder, monkeypatch):
+# locale is set in the bootloader during python interpreter setup). In its current form, the test is mostly ensuring
+# that the user-preferred locale is set *at all* (see #8305).
+#
+# NOTE: monkey-patching LC_ALL might cause issues during build. Ideally, we would modify the LC_ALL only when running
+# built executable, but currently `pyi_builder.test_source` does not allow propagation of run-time environment
+# variables. If that is ever fixed/extended, the test could include a wider range of locale combinations to test with
+# (for example, sl_SI.UTF-8, sl_SI.ISO8859-2).
```
We do already have a few tests which fish out the executable and `subprocess.run()` it. You could do that?

rokm replied:
Hmm, true. We could pass the expected locale and encoding via command-line arguments, and then use a single build to test multiple variants. Without arguments, it would just check that values are not `None`, and with arguments, it would check that they are what we expect them to be.

rokm commented:
Hmm, looks like the CI runner does not come with the extra locales. Should we adjust the workflow to install/enable them, or can we live without that part of the test?

rokm committed:
ci: install/enable extra locales on ubuntu CI runner

The ubuntu CI runner does not come with any locales enabled, so
we need to explicitly enable the following locales for
`test_basic::test_user_preferred_locale`:
 - en_US.UTF-8
 - en_US (= en_US.ISO8859-1)
 - sl_SI.UTF-8
 - sl_SI (= sl_SI.ISO8859-2)

In addition, `test_basic::test_time_module_localized` seems to
be using `cs_CZ.UTF-8`, so enable that as well.

bwoodsend commented:
I see you've already done it but yeah, I think installing the locales on CI sounds sensible.
